Create unlinked reference doc for new file system based route APIs (#26222)
gatsbyjs/gatsby: docs/docs/file-system-page-creation.md

---
title: Creating Pages from the File System
---

This page roughly documents the APIs and conventions available with the new file-based page creation features.

For now, these features are marked as experimental and require a flag to utilize. Start your Gatsby site locally with the following flag:

```bash
GATSBY_EXPERIMENTAL_ROUTING_APIS=1 gatsby develop
```

## Creating client-only routes

Client-only routes can be created by using square brackets (`[ ]`) in the file path to mark any dynamic segments of the URL. While Gatsby creates predictable routes for static pages, the same expectations apply to client side routes. Here are some examples:

- `/src/pages/users/[id].js => /users/:id`
- `/src/pages/products/[...splat].js => /users/*splat`
- `/src/pages/app/[...].js => /app/*`

> The dynamic segment of the file name (the part between the square brackets) will be filled in and provided to your components on a `props.params` object. For example:

Reviewer: Could you expand a little bit on the syntax here and lay out what `...` and `*` are for in this context? I'm guessing not everyone who wants to use this will be familiar.

Author: I have updated it, wanna check it out?

```js:title=/src/pages/users/[id].js
function UserPage(props) {
  const id = props.params.id
}
```

```js:title=/src/pages/products/[...splat].js
function ProductsPage(props) {
  const splat = props.params.splat
}
```

```js:title=/src/pages/app/[...].js
function AppPage(props) {
  const splat = props.params[‘*’]
}
```

## Creating collection pages

Developers can create multiple pages from a model based on the collection of nodes within it. To do that, use curly braces (`{ }`) to signify dynamic URL segments that relate to a field within the node. There is some special logic that can happen in here. Here are a few examples:

- `/src/pages/products/{Product.name}.js => /products/burger`
- `/src/pages/products/{Product.fields__sku}.js => /products/001923`
- `/src/pages/blog/{MarkdownRemark.parent__(File)__relativePath}.js => /blog/learning-gatsby`

> We use the content within the curly braces to generate GraphQL queries to retrieve the nodes that should be built for a given collection. For example:

Reviewer: ```suggestion
Gatsby uses the content within the curly braces to generate GraphQL queries to retrieve the nodes that should be built for a given collection. For example:
```

```jsx:title=/src/pages/products/{Product.name}.js
import { unstable_collectionGraphql } from "gatsby"

export default function Component(props) {
  return props.data.fields.sku + props.params.name
}

// If you are customizing the collection query, there is a special fragment you MUST use when using this API. The fragment converts to
// { nodes { id, [params_from_path] } }
export const collectionQuery = unstable_collectionGraphql`
{
  allProduct(filter: { type: { nin: ["Burger"] } }) {
    ...CollectionPagesQueryFragment
  }
}`

// This is the page query that connects the data to the actual component. Here you can query for any and all fields
// you need access to within your code. Again, since Gatsby always queries for `id` in the collection, you can use that
// to connect to this GraphQL query.
export const query = graphql`
  query ($id: String) {
    product(id: { eq: $id }) {
      fields {
        sku
      }
    }
  }
}
```

## Routing and linking

> We "slugify" every route that is created from collection pages. When you want to link to one of those pages, it may not always be clear how to construct the URL from scratch.

Reviewer: ```suggestion
Gatsby "slugifies" every route that gets created from collection pages. When you want to link to one of those pages, it may not always be clear how to construct the URL from scratch.
```

---
title: Creating Pages from the File System
---

This page roughly documents the APIs and conventions available with the new file-based page creation features.

For now, these features are marked as experimental and require a flag to utilize. Start your Gatsby site locally with the following flag:

> ```bash

Reviewer 2: ````suggestion
```shell
````

GATSBY_EXPERIMENTAL_ROUTING_APIS=1 gatsby develop
```

## Creating client-only routes

Use client-only routes if you have dynamic data that does not live in Gatsby. This might be something like a user settings page, or some other dynamic content that isn't known to Gatsby at build time. In these situations, you will usually create a route with one or more dynamic segments to query data from a server in order to render your page.

For example, in order to edit a user, you might want a route like `/user/:id` to fetch the data for whatever `id` is passed into the URL. You can now use square brackets (`[ ]`) in the file path to mark any dynamic segments of the URL.

- `/src/pages/users/[id].js => /users/:id`
- `/src/pages/users/[id]/group/[groupId].js => /users/:id/group/:groupId`

Gatsby also supports _splat_ routes, which are routes that will match _anything_ after the splat. These are less common, but still have use cases. As an example, suppose that you are rendering images from [S3](/docs/deploying-to-s3-cloudfront/) and the URL is actually the key to the asset in AWS. Here is how you might create your file:

- `/src/pages/image/[...awsKey].js => /users/*awsKey`

> - `/src/pages/image/[...].js => /app/*`

Reviewer 3: the right side doesn't match left side here - on the left `/images/splat` - on the right `/users/splat` or `/app/splat`

## Creating client-only routes

Client-only routes are pages that would be used if you have dynamic data that does not live in Gatsby. This might be something like a user settings page, or some other dynamic content that isn't known to Gatsby at build time. In these situations you usually create a route with dynamic segment(s) in which you use to query data from a server to render your page.

> For example, if you are editing a user you might want a route like `/user/:id` to then fetch the data for whatever `id` is passed in to the URL. Client only pages are now quite simple to create with a new file system convention. Use square brackets (`[ ]`) in the file path to mark any dynamic segments of the URL.

Reviewer: ```suggestion
For example, in order to edit a user, you might want a route like `/user/:id` to fetch the data for whatever `id` is passed into the URL. You can now use square brackets (`[ ]`) in the file path to mark any dynamic segments of the URL.
```

- `/src/pages/users/[id].js => /users/:id`
- `/src/pages/users/[id]/group/[groupId].js => /users/:id/group/:groupId`

> We also support _splat_ routes, which are routes that will match _anything_ after the splat. These are less common, but still have use-cases. One example might be that you are rendering images from S3, and the url is actually the key to the asset in AWS. Here is how you might create your file:

Reviewer: ```suggestion
Gatsby also supports _splat_ routes, which are routes that will match _anything_ after the splat. These are less common, but still have use cases. As an example, suppose that you are rendering images from [S3](/docs/deploying-to-s3-cloudfront/) and the URL is actually the key to the asset in AWS. Here is how you might create your file:
```

- `/src/pages/image/[...awsKey].js => /users/*awsKey`
- `/src/pages/image/[...].js => /app/*`

> Three periods `...` mark a page as a splat route. You can optionally name the splat as well, which has the benefit of naming the key of the property that your component receives. The dynamic segment of the file name (the part between the square brackets) will be filled in and provided to your components on a `props.params` object. For example:

Reviewer: ```suggestion
Three periods `...` mark a page as a splat route. Optionally, you can name the splat as well, which has the benefit of naming the key of the property that your component receives. The dynamic segment of the file name (the part between the square brackets) will be filled in and provided to your components on a `props.params` object. For example:
```
